docs/source/faq.rst

```diff
@@ -110,5 +110,239 @@ Based on `this blog article <https://medium.com/huggingface/training-larger-batc
+
+    @trainer.on(Events.ITERATION_COMPLETED(once=15))
+    def stop_training():
+        trainer.terminate()
+
+    trainer.run(infinite_iterator(4))
+
+We obtain the following output:
+
+.. code-block:: text
+
+    1/1 : 1 - 63.862
+    1/1 : 2 - 64.042
+    1/1 : 3 - 63.936
+    1/1 : 4 - 64.141
+    1/1 : 5 - 64.767
+    1/1 : 6 - 63.791
+    1/1 : 7 - 64.565
+    1/1 : 8 - 63.602
+    1/1 : 9 - 63.995
+    1/1 : 10 - 63.943
+    1/1 : 11 - 63.831
+    1/1 : 12 - 64.276
+    1/1 : 13 - 64.148
+    1/1 : 14 - 63.920
+    1/1 : 15 - 64.226
+
+
+Same code can be used for validating models.
+
+
+Finite iterator with unknown length
+```````````````````````````````````
+
+Let's use a finite data iterator but with unknown length (for user). In case of training, we would like to perform
+several passes over the dataflow and thus we need to restart the data iterator when it is exhausted.
+In the code, we do not specify `epoch_length` which will be automatically determined.
+
+.. code-block:: python
+
+    import torch
+    from ignite.engine import Engine, Events
+
+    torch.manual_seed(12)
+
+    def finite_unk_size_data_iter():
+        for i in range(11):
+            yield i
+
+    def train_step(trainer, batch):
+        # ...
+        s = trainer.state
+        print(
+            "{}/{} : {} - {:.3f}".format(s.epoch, s.max_epochs, s.iteration, batch)
+        )
+
+    trainer = Engine(train_step)
+
+    @trainer.on(Events.DATALOADER_STOP_ITERATION)
+    def restart_iter():
+        trainer.state.dataloader = finite_unk_size_data_iter()
+
+    data_iter = finite_unk_size_data_iter()
+    trainer.run(data_iter, max_epochs=5)
+
+
+In case of validation, the code is simply
+
+.. code-block:: python
+
+    import torch
+    from ignite.engine import Engine, Events
+
+    torch.manual_seed(12)
+
+    def finite_unk_size_data_iter():
+        for i in range(11):
+            yield i
+
+    def val_step(evaluator, batch):
+        # ...
+        s = evaluator.state
+        print(
+            "{}/{} : {} - {:.3f}".format(s.epoch, s.max_epochs, s.iteration, batch)
+        )
+
+    evaluator = Engine(val_step)
+
+    data_iter = finite_unk_size_data_iter()
+    evaluator.run(data_iter)
+
+
+Finite iterator with known length
+`````````````````````````````````
+
+Let's use a finite data iterator with known size for training or validation.
+If we need to restart the data iterator, we can do this explicitly on iteration.
+
+.. code-block:: python
+
+    import torch
+    from ignite.engine import Engine, Events
+
+    torch.manual_seed(12)
+
+    size = 11
+
+    def finite_size_data_iter(size):
+        for i in range(size):
+            yield i
+
+    def train_step(trainer, batch):
+        # ...
+        s = trainer.state
+        print(
+            "{}/{} : {} - {:.3f}".format(s.epoch, s.max_epochs, s.iteration, batch)
+        )
+
+    trainer = Engine(train_step)
+
+    @trainer.on(Events.ITERATION_COMPLETED(every=size))
```

Comment:
It could be the code of the previous section (using DATALOADER_STOP_ITERATION) ? 

Is it possible to restart iterator in backend using `epoch_lenght` for example ?

Btw, API is clear and simple.

Reply:
> It could be the code of the previous section (using DATALOADER_STOP_ITERATION) ?

yes, it can be, but here we show that as we know the lenght we can handle it explicitly.
I'll update the docs.

> Is it possible to restart iterator in backend using epoch_lenght for example ?

In general case, we do not know how to restart an iterator.

Reply:
I think it is not mandatory to update doc, cases are well splitted IMO :)